docs/proposals/202104_native_support_of_multitenancy.md

```diff
@@ -0,0 +1,83 @@
+    * Evaluate some set of rules and check if a certain label can access the data or not.
+
+### Solution
+
+* Introduce a new flag to enable or disable the multi-tenancy mode.
+* When this flag is enabled we will modify the metrics which can be used to track usage characteristics of a tenant by injecting the tenant’s label (a const. Label ).
+* We will make sure that the query request received has a tenant label and if not we have some default tenant label configured.
+* We will enforce the tenant labels inside the Thanos code base(i.e we assume the authentication is done and the query has access to the tenant's data which it specifies) when we run components with a specified mode(i.e multi-tenancy flag is enabled). We will no longer require any label enforcing proxy.
+* Support the cross tenant queries using regular expressions.
+* Store API will advertise the tenant label and this how it will help us -
+    * Let's assume a case of hard tenancy, we have two store-gateway advertising different sets of tenants.
+    * Store-gateway - 1 advertises tenant-a, tenant-b.
+    * Store-gateway - 2 advertises tenant-c, tenant-d.
+    * For the query, we will have both the store-gateway, and depending on the query it will go to the desired store-gateway. Same goes for receive.
+* So, when we get a query and we pass it, and we have a tenant label and we treat that as a special value for a tenant, we can extract that value and aggregate some metrics and inject these metrics into Thanos components itself. Then we have tenant-specific metrics. Depending on the data we collect from the Thanos we can generate the dashboard etc.
+* Changes specific to components
+    * Ruler
+        * We have put some extra effort into Ruler, there will be rules according to different tenants and there will be a configuration layer from tenant to which alert manager(i.e for different tenants having different alert manager ) to redirect. We have tenant-id(tenant label) and we can inject them on all outgoing requests as header and querier knows how to deal with them, we get the queries done and if we get some alert fired, we will redirect to tenant-specific alert managers.
+    * Store
+        * We will build multi-tenancy on a label-based approach and after introducing this PR we will need a thin layer that will convert custom prefix to external label and the rest advertising label, querying will be the same.
+
+#### Alternatives
+
+1. If we want to track the resource usages of a query, we can deploy another X proxy(hypothetical) depending on the data we want to aggregate. Similar to prom-label proxy we can inject this proxy to each of thanos components and check the queries and depending on tenant label(or any specified label) we can aggregate the metrics, this would be okay and we can aggregate simple metrics like error rates, latencies. And then we generate metrics per tenant and scrape these proxies from another prometheus and then finally generate dashboard reports etc.
```

Comment:
```suggestion
1. If we want to track the resource usages of a query, we can deploy another X proxy (hypothetical) depending on the data we want to aggregate. Similar to prom-label proxy we can inject this proxy to each of thanos components and check the queries and depending on tenant label (or any specified label) we can aggregate the metrics, this would be okay and we can aggregate simple metrics like error rates, latencies. And then we generate metrics per tenant and scrape these proxies from another prometheus and then finally generate dashboard reports etc.
```

Reply:
addressed